packages/grid/x-data-grid-generator/src/hooks/useQuery.ts

```diff
@@ -0,0 +1,280 @@
+        const value = valueGetters[index](row);
+        return filterFunctions[index] === null ? true : filterFunctions[index]({ value });
+      }),
+    );
+  }
+  return rows.filter((row: GridRowModel) =>
+    filterModel.items.every((_, index) => {
+      const value = valueGetters[index](row);
+      return filterFunctions[index] === null ? true : filterFunctions[index]({ value });
+    }),
+  );
+};
+
+/**
+ * Simulates server data loading
+ */
+const loadServerRows = (
+  rows: GridRowModel[],
+  queryOptions: QueryOptions,
+  serverOptions: ServerOptions,
+  columnsWithDefaultColDef: GridColDef[],
+): Promise<FakeServerResponse> => {
+  const { minDelay = 100, maxDelay = 300, useCursorPagination } = serverOptions;
+
+  if (maxDelay < minDelay) {
+    throw new Error('serverOptions.minDelay is larger than serverOptions.maxDelay ');
+  }
+  const delay = Math.random() * (maxDelay - minDelay) + minDelay;
+
+  const { cursor, page = 0, pageSize } = queryOptions;
+
+  let nextCursor;
+  let firstRowIndex;
+  let lastRowIndex;
+
+  let filteredRows = getFilteredRows(rows, queryOptions.filterModel, columnsWithDefaultColDef);
+
+  const rowComparator = getRowComparator(queryOptions.sortModel, columnsWithDefaultColDef);
+  filteredRows = [...filteredRows].sort(rowComparator);
+
+  const totalRowCount = filteredRows.length;
+  if (!pageSize) {
+    firstRowIndex = 0;
+    lastRowIndex = filteredRows.length;
+  } else if (useCursorPagination) {
+    firstRowIndex = cursor ? filteredRows.findIndex(({ id }) => id === cursor) : 0;
+    firstRowIndex = Math.max(firstRowIndex, 0); // if cursor not found return 0
+    lastRowIndex = firstRowIndex + pageSize;
+
+    nextCursor = lastRowIndex >= filteredRows.length ? undefined : filteredRows[lastRowIndex].id;
+  } else {
+    firstRowIndex = page * pageSize;
+    lastRowIndex = (page + 1) * pageSize;
+  }
+  const response: FakeServerResponse = {
+    returnedRows: filteredRows.slice(firstRowIndex, lastRowIndex),
+    nextCursor,
+    totalRowCount,
+  };
+
+  return new Promise<FakeServerResponse>((resolve) => {
+    setTimeout(() => {
+      resolve(response);
+    }, delay); // simulate network latency
+  });
+};
+
+interface FakeServerResponse {
+  returnedRows: GridRowModel[];
+  nextCursor?: string;
+  totalRowCount: number;
+}
+
+interface PageInfo {
+  totalRowCount?: number;
+  nextCursor?: string;
+  pageSize?: number;
+}
+
+export interface ServerOptions {
+  minDelay: number;
+  maxDelay: number;
+  useCursorPagination?: boolean;
+}
+
+export interface QueryOptions {
+  cursor?: GridRowId;
+  page?: number;
+  pageSize?: number;
+  // TODO: implement the behavior liked to following models
+  filterModel?: GridFilterModel;
+  sortModel?: GridSortModel;
+}
+
+const DEFAULT_DATASET_OPTIONS: UseDemoDataOptions = {
+  dataSet: 'Commodity',
+  rowLength: 100,
+  maxColumns: 6,
+};
+
+const DEFAULT_SERVER_OPTIONS: ServerOptions = {
+  minDelay: 100,
+  maxDelay: 300,
```

Comment:
We might want to add a logic as we did for `DISABLE_CHANCE_RANDOM` to disable this for the visual regression tests, it's unstable: https://www.argos-ci.com/mui/mui-x/builds/1826

<img width="1023" alt="Screenshot 2022-06-05 at 15 37 53" src="https://user-images.githubusercontent.com/3165635/172053518-4d1e2c8c-9963-4358-9e7a-d2db4ab37965.png">